pkgs/dart_tooling_mcp_server/test/test_harness.dart

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:async/async.dart';
+import 'package:dart_mcp/client.dart';
+import 'package:dart_tooling_mcp_server/src/mixins/dtd.dart';
+import 'package:dart_tooling_mcp_server/src/server.dart';
+import 'package:devtools_shared/devtools_shared.dart';
+import 'package:dtd/dtd.dart';
+import 'package:http/http.dart' as http;
+import 'package:stream_channel/stream_channel.dart';
+import 'package:test/test.dart';
+import 'package:test_descriptor/test_descriptor.dart' as d;
+import 'package:test_process/test_process.dart';
+
+/// A full environment for integration testing the MCP server.
+///
+/// - Runs the counter app at `test_fixtures/counter_app` using `flutter run`.
+/// - Connects to the dtd service and registers a fake `Editor.getDebugSessions`
+///   extension method on it to mimic the DartCode extension.
+/// - Sets up the MCP client and server, and fully initializes the connection
+///   between them. Includes a debug mode for running them both in process to
+///   allow for breakpoints, but the default mode is to run the server in a
+///   separate process.
+class TestHarness {
+  final TestProcess flutterProcess;
+  final DartToolingMCPClient mcpClient;
+  final ServerConnection mcpServerConnection;
+  final String dtdUri;
+
+  TestHarness._(
+    this.flutterProcess,
+    this.mcpClient,
+    this.mcpServerConnection,
+    this.dtdUri,
+  );
+
+  /// Starts a flutter process as well as an MCP client and server.
+  ///
+  /// Handles the initialization handshake between the MCP client and server as
+  /// well.
+  ///
+  /// By default this will run with the MCP server compiled as a separate binary
+  /// to mimic as closely as possible the real world behavior. This makes it so
+  /// breakpoints in the server don't work however, so you can set [debugMode]
+  /// to `true` and we will run it in process instead, which allows breakpoints
+  /// since everything is running in the same isolate.
+  static Future<TestHarness> start({
+    @Deprecated('For debugging only, do not submit') bool debugMode = false,
+  }) async {
+    final platform = Platform.isLinux
+        ? 'linux'
+        : Platform.isMacOS
+            ? 'macos'
+            : throw StateError(
+                'unsupported platform, only mac and linux are supported',
+              );
+    final flutterProcess = await TestProcess.start(
+      // TODO: Get flutter SDK location from somewhere.
```

Comment:
you could look at the devtools ci integration for inspiration WRT checking out flutter and knowing where the flutter sdk is

Reply:
This is quite a lot simpler right now and I would generally like to keep it that way - I will file an issue for this TODO though.